compiler/test/dotty/tools/dotc/transform/SpecializeFunction1Tests.scala

```diff
@@ -0,0 +1,94 @@
+package dotty.tools
+package dotc
+package transform
+
+import org.junit.Assert._
+import org.junit.Test
+
+import dotty.tools.backend.jvm.DottyBytecodeTest
+
+class SpecializeFunction1Tests extends DottyBytecodeTest {
+
+  import dotty.tools.backend.jvm.ASMConverters._
+  import dotty.tools.backend.jvm.AsmNode._
+
+  protected def checkForBoxing(ins: List[Instruction], source: String): Unit = ins.foreach {
+    case Invoke(op, owner, name, desc, itf) =>
+      def error =
+        s"""|----------------------------------
+            |${ins.mkString("\n")}
+            |----------------------------------
+            |From code:
+            |$source
+            |----------------------------------""".stripMargin
+
+      assert(!owner.toLowerCase.contains("box"), s"Boxing instruction discovered in:\n$error")
+      assert(!name.toLowerCase.contains("box"), s"Boxing instruction discovered in:\n$error")
+    case _ => ()
+  }
+
+  @Test def specializeParentIntToInt = {
+    val source = """
+                 |class Foo extends Function1[Int, Int] {
+                 |  def apply(i: Int) = i
+                 |}
+                 """.stripMargin
+
+    checkBCode(source) { dir =>
+      import scala.collection.JavaConverters._
+      val clsIn = dir.lookupName("Foo.class", directory = false).input
+      val clsNode = loadClassNode(clsIn)
+      assert(clsNode.name == "Foo", s"inspecting wrong class: ${clsNode.name}")
+      val methods = clsNode.methods.asScala
+
+      val applys = methods
+        .collect {
+          case m if m.name == "apply$mcII$sp" => m
+          case m if m.name == "apply" => m
+        }
+        .map(_.name)
+        .toList
+
+      assert(
```

Comment:
it would be also nice to check that the actual implementation is in the right apply.
Ie, `apply$mcII$sp` isn't a forwarder to generic `apply` that unboxes.